lib/fetcher.js

```diff
@@ -351,8 +351,13 @@ class FetcherBase {
       ? mkdirp(dir).then(() => this[_toFile](dest))
       : inferOwner(dest).then(({uid, gid}) =>
         mkdirp(dir).then(made => this[_toFile](dest)
-          .then(res => this[_chown](made || dir, uid, gid)
-            .then(() => res))))
+          .then(res => {
+            const p = this[_chown](made || dir, uid, gid)
+            return p
+              ? p.then(() => res)
+              : /* istanbul ignore next - we don't test in root-owned folders */
+                Promise.resolve(res)
```

Comment:
This can just be `return p ? p.then(() => res) : res`, because we're already in a Promise chain.

Reply:
That is, I think your initial way of doing it was correct, just needed the `/* istanbul ignore next */` comment.

Reply:
or, better yet, maybe just make the `[_chown]` function async?  No reason to _sometimes_ return a Promise, after all.  Seems like this solves the problem as well, without needing to duplicate the istanbul override:

```patch
diff --git a/lib/fetcher.js b/lib/fetcher.js
index ba8057b..33fbf79 100644
--- a/lib/fetcher.js
+++ b/lib/fetcher.js
@@ -295,7 +295,7 @@ class FetcherBase {
     return cacache.rm.content(this.cache, this.integrity, this.opts)
   }
 
-  [_chown] (path, uid, gid) {
+  async [_chown] (path, uid, gid) {
     return selfOwner && (selfOwner.gid !== gid || selfOwner.uid !== uid)
       ? chownr(path, uid, gid)
       : /* istanbul ignore next - we don't test in root-owned folders */ null
```

Reply:
Today I learned - this is much nicer! Thank you for the improvements, and for making the edits.